File: sdk/typescript/src/namespaces/laboratory.ts
/**
 * Laboratory Namespace API
 *
 * Provides access to the persona laboratory for detecting emergent traits
 * and suggesting beneficial cross-pollinations between agents.
 */

/**
 * An emergent trait detected from agent performance.
 */
export interface EmergentTrait {
  agent: string;
  trait: string;
  domain: string;
  confidence: number;
  evidence: string[];
  detected_at: string;
}

/**
 * A suggested trait transfer between agents.
 */
export interface CrossPollinationSuggestion {
  source_agent: string;
  trait_or_domain: string;
  reason: string;
}

/**
 * Response for emergent traits listing.
 */
export interface EmergentTraitsResponse {
  emergent_traits: EmergentTrait[];
  count: number;
  min_confidence: number;
}

/**
 * Response for cross-pollination suggestions.
 */
export interface CrossPollinationResponse {
  target_agent: string;
  suggestions: CrossPollinationSuggestion[];
  count: number;
}

/**
 * Options for listing emergent traits.
 */
export interface EmergentTraitsOptions {
  min_confidence?: number;
  limit?: number;
}

/**
 * Request for cross-pollination suggestions.
 */
export interface CrossPollinationRequest {
  target_agent: string;
}

/**
 * Interface for the internal client methods used by LaboratoryAPI.
 */
interface LaboratoryClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Laboratory API namespace.
 *
 * Provides access to persona laboratory features for detecting
 * emergent traits and suggesting beneficial cross-pollinations.
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Get emergent traits with high confidence
 * const traits = await client.laboratory.getEmergentTraits({
 *   min_confidence: 0.8,
 *   limit: 10
 * });
 * for (const trait of traits.emergent_traits) {
 *   console.log(`${trait.agent}: ${trait.trait} (${trait.confidence})`);
 * }
 *
 * // Get cross-pollination suggestions
 * const suggestions = await client.laboratory.suggestCrossPollinations({
 *   target_agent: 'claude'
 * });
 * for (const s of suggestions.suggestions) {
 *   console.log(`Transfer ${s.trait_or_domain} from ${s.source_agent}`);
 * }
 * ```
 */
export class LaboratoryAPI {
  constructor(private client: LaboratoryClientInterface) {}

  /**
   * Get emergent traits detected from agent performance.
   */
  async getEmergentTraits(options?: EmergentTraitsOptions): Promise<EmergentTraitsResponse> {
    return this.client.request('GET', '/api/v2/laboratory/emergent-traits', {
      params: options as Record<string, unknown>,
    });
  }

  /**
   * Suggest beneficial trait transfers for a target agent.
   */
  async suggestCrossPollinations(request: CrossPollinationRequest): Promise<CrossPollinationResponse> {
    return this.client.request('POST', '/api/v2/laboratory/cross-pollinations', {
      json: request as unknown as Record<string, unknown>,
    });
  }
}

File: sdk/python/aragora_sdk/namespaces/laboratory.py
"""
Laboratory Namespace API.

Provides access to the persona laboratory for detecting emergent traits
and suggesting beneficial cross-pollinations between agents.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class LaboratoryAPI:
    """
    Synchronous Laboratory API.

    Provides access to persona laboratory features for detecting
    emergent traits and suggesting beneficial cross-pollinations.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> traits = client.laboratory.get_emergent_traits(min_confidence=0.8)
        >>> suggestions = client.laboratory.suggest_cross_pollinations("claude")
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    def get_emergent_traits(
        self,
        min_confidence: float | None = None,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """
        Get emergent traits detected from agent performance.

        Args:
            min_confidence: Minimum confidence threshold (0.0-1.0).
            limit: Maximum number of traits to return.

        Returns:
            List of emergent traits with confidence scores and evidence.
        """
        params: dict[str, Any] = {}
        if min_confidence is not None:
            params["min_confidence"] = min_confidence
        if limit is not None:
            params["limit"] = limit

        return self._client._request("GET", "/api/v2/laboratory/emergent-traits", params=params)

    def suggest_cross_pollinations(
        self,
        target_agent: str,
    ) -> dict[str, Any]:
        """
        Suggest beneficial trait transfers for a target agent.

        Args:
            target_agent: Agent to receive trait suggestions.

        Returns:
            Cross-pollination suggestions with source agents and reasons.
        """
        return self._client._request(
            "POST",
            "/api/v2/laboratory/cross-pollinations",
            json={"target_agent": target_agent},
        )

    def get_trait_analysis(
        self,
        agent_name: str,
    ) -> dict[str, Any]:
        """
        Get detailed trait analysis for an agent.

        Args:
            agent_name: Name of the agent to analyze.

        Returns:
            Trait analysis with strengths, weaknesses, and recommendations.
        """
        return self._client._request("GET", f"/api/v2/laboratory/agent/{agent_name}/analysis")

    def run_experiment(
        self,
        experiment_type: str,
        agents: list[str],
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Run a laboratory experiment.

        Args:
            experiment_type: Type of experiment (trait_transfer, comparison, etc.).
            agents: List of agents to include in experiment.
            config: Optional experiment configuration.

        Returns:
            Experiment results with metrics and observations.
        """
        data: dict[str, Any] = {
            "experiment_type": experiment_type,
            "agents": agents,
        }
        if config:
            data["config"] = config

        return self._client._request("POST", "/api/v2/laboratory/experiments", json=data)

    def get_experiments(
        self,
        limit: int | None = None,
        status: str | None = None,
    ) -> dict[str, Any]:
        """
        List laboratory experiments.

        Args:
            limit: Maximum number of experiments to return.
            status: Filter by status (pending, running, completed).

        Returns:
            List of experiments with status and results.
        """
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if status is not None:
            params["status"] = status

        return self._client._request("GET", "/api/v2/laboratory/experiments", params=params)


class AsyncLaboratoryAPI:
    """Asynchronous Laboratory API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    async def get_emergent_traits(
        self,
        min_confidence: float | None = None,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """Get emergent traits detected from agent performance."""
        params: dict[str, Any] = {}
        if min_confidence is not None:
            params["min_confidence"] = min_confidence
        if limit is not None:
            params["limit"] = limit

        return await self._client._request(
            "GET", "/api/v2/laboratory/emergent-traits", params=params
        )

    async def suggest_cross_pollinations(
        self,
        target_agent: str,
    ) -> dict[str, Any]:
        """Suggest beneficial trait transfers for a target agent."""
        return await self._client._request(
            "POST",
            "/api/v2/laboratory/cross-pollinations",
            json={"target_agent": target_agent},
        )

    async def get_trait_analysis(
        self,
        agent_name: str,
    ) -> dict[str, Any]:
        """Get detailed trait analysis for an agent."""
        return await self._client._request("GET", f"/api/v2/laboratory/agent/{agent_name}/analysis")

    async def run_experiment(
        self,
        experiment_type: str,
        agents: list[str],
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Run a laboratory experiment."""
        data: dict[str, Any] = {
            "experiment_type": experiment_type,
            "agents": agents,
        }
        if config:
            data["config"] = config

        return await self._client._request("POST", "/api/v2/laboratory/experiments", json=data)

    async def get_experiments(
        self,
        limit: int | None = None,
        status: str | None = None,
    ) -> dict[str, Any]:
        """List laboratory experiments."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if status is not None:
            params["status"] = status

        return await self._client._request("GET", "/api/v2/laboratory/experiments", params=params)
